errors.go

```diff
@@ -9,6 +9,21 @@ import (
 	"strconv"
 )
 
+// BatchError is an interface that extends the builtin error interface with an
+// array of errors.
+type BatchError interface {
+	error
+	Errors() []error
+	IsEmpty() bool
+	Append(error)
```

Comment:
We should re-think whether we need `IsEmpty` and `Append`. We could have a very small interface.